### PR TITLE
Fix add_peer_test

### PR DESCRIPTION
**Pull request by kuvaldini**
- Fix race condition in add_peer_test
- small refactotign
- [refac] report_abort() to protp_tx_resp.cpp

### CODE REVIEW

File: shared_model/backend/protobuf/transaction_responses/impl/proto_tx_response.cpp [hunk 1]
```diff
@@ -45,26 +39,31 @@ namespace shared_model::proto {
...
+      using namespace shared_model::proto;
+      switch (proto_.tx_status()) {
+        case iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED:
+          return ProtoResponseVariantType(StatelessFailedTxResponse(proto_));
+        case iroha::protocol::TxStatus::STATELESS_VALIDATION_SUCCESS:
+          return ProtoResponseVariantType(StatelessValidTxResponse(proto_));
+        case iroha::protocol::TxStatus::STATEFUL_VALIDATION_FAILED:
+          return ProtoResponseVariantType(StatefulFailedTxResponse(proto_));
+        case iroha::protocol::TxStatus::STATEFUL_VALIDATION_SUCCESS:
+          return ProtoResponseVariantType(StatefulValidTxResponse(proto_));
```
Comment (iceseer) [inline shared_model/backend/protobuf/transaction_responses/impl/proto_tx_response.cpp:51]: make more compact plz

File: test/framework/integration_framework/integration_test_framework.hpp [hunk 2]
```diff
@@ -506,8 +506,31 @@ namespace integration_framework {
...
+    struct ResponsesQueues;
+    //    {
+    //     public:
+    //      //using HashType = shared_model::crypto::Hash;
+    //      std::map<std::string, std::unique_ptr<CheckerQueue<TxResponseType>>>
+    //      map;
+    //     private:
+    //      /// maximum time of waiting before appearing next transaction
+    //      response std::chrono::milliseconds tx_response_waiting_time_ms;
+    //      std::recursive_mutex mtx;
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.hpp:519]: remove plz

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 3]
```diff
@@ -133,26 +129,103 @@ class IntegrationTestFramework::CheckerQueue {
...
+                     std::unique_ptr<CheckerQueue<TxResponseType>>,
+                     HashType::Hasher>
+      map;
+
+ public:
+  ResponsesQueues(std::chrono::milliseconds ms)
+      : tx_response_waiting_time_ms(ms) {}
+  auto lock() {
+    return std::unique_lock(mtx);
+  }
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:177]: bad design, don't do such a way

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 4]
```diff
@@ -133,26 +129,103 @@ class IntegrationTestFramework::CheckerQueue {
...
+    return it;
+  }
+  auto find(HashType const &txhash) -> std::optional<decltype(map)::iterator> {
+    // assert(not mtx.try_lock());  // expecting it is locked before
+    auto lk = lock();
+    auto it = map.find(txhash);
+    if (it == map.end()) {
+      return std::nullopt;
+    }
+    return {it};
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:197]: potential bug

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 5]
```diff
@@ -133,26 +129,103 @@ class IntegrationTestFramework::CheckerQueue {
...
+#include <unordered_map>
+
+struct IntegrationTestFramework::ResponsesQueues {
+ public:
+  using HashType = shared_model::interface::types::HashType;
+
+ private:
+  /// maximum time of waiting before appearing next transaction response
+  std::chrono::milliseconds tx_response_waiting_time_ms;
+  std::recursive_mutex mtx;
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:166]: You don't need recursive mutex here. It is a bad way to lock outside and enter under mutex twice

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 6]
```diff
@@ -190,7 +266,7 @@ IntegrationTestFramework::IntegrationTestFramework(
               kLocalHost, torii_port_, std::nullopt))),
       async_call_(std::make_shared<AsyncCall>(
           log_manager_->getChild("AsyncCall")->getLogger())),
-      tx_response_waiting(tx_response_waiting),
+      //      tx_response_waiting_time_ms_(tx_response_waiting),
```
Comment (kuvaldini) [inline test/framework/integration_framework/integration_test_framework.cpp:269]:
```suggestion
```

File: shared_model/backend/protobuf/transaction_responses/impl/proto_tx_response.cpp (hunk 1 above)
Comment (kuvaldini) [reply shared_model/backend/protobuf/transaction_responses/impl/proto_tx_response.cpp]: done

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 3 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.cpp]: remade

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 4 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.cpp]: reworked

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 5 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.cpp]: reworked

File: test/framework/integration_framework/integration_test_framework.hpp (hunk 2 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.hpp]: done

File: irohad/main/server_runner.cpp [hunk 7]
```diff
@@ -77,7 +75,7 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
 
   if (selected_port == 0) {
     return iroha::expected::makeError(
-        (boost::format(kPortBindError) % server_address_).str());
+        (boost::format("Cannot bind server to address %s") % server_address_).str());
```
Comment (iceseer) [inline irohad/main/server_runner.cpp:78]: fmt::format ?

File: test/framework/integration_framework/fake_peer/fake_peer.cpp [hunk 8]
```diff
@@ -212,6 +212,8 @@ FakePeer::getProposalStorage() {
 std::unique_ptr<iroha::network::ServerRunner> FakePeer::run() {
   ensureInitialized();
   // start instance
+  using namespace std::chrono_literals;
+  std::this_thread::sleep_for(500ms);
```
Comment (iceseer) [inline test/framework/integration_framework/fake_peer/fake_peer.cpp:216]: Seems like a hack. Plz insert TODO to wait untill server stop and all handlers will be joined

File: test/framework/integration_framework/fake_peer/fake_peer.hpp [hunk 9]
```diff
@@ -40,6 +40,13 @@ namespace integration_framework::fake_peer {
             shared_model::interface::Transaction,
             iroha::protocol::Transaction>;
 
+    /** Instead of constructor because shared_from_this
+     */
+    template<class...Args>
+    static std::shared_ptr<FakePeer> create_shared(Args&&...args){
```
Comment (iceseer) [inline test/framework/integration_framework/fake_peer/fake_peer.hpp:46]: createShared

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 10]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+      -> WaitGetResult {  // std::tuple<TxResponsePtr,
+                          // std::chrono::milliseconds> {  //
+                          // std::optional<TxResponsePtr>
+                          //  {
+    std::unique_lock lk(mtx);
+    auto it = map.find(txhash);
+    if (it == map.end())
+      return {nullptr, 0ms};
+    auto &qu = it->second;
+    auto const deadline = std::chrono::steady_clock::now() + timeout;
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:195]:
```c++
auto const measure_start = std::chrono::steady_clock::now();
auto const deadline = measure_start + timeout;
...
auto const elapsed = td::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now() - measure_start);

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 11]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+        ;
+    }
+    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+        deadline - std::chrono::steady_clock::now());
+    if (qu.empty())  // timed out and still empty
+      return {nullptr, elapsed};
+    if constexpr (do_pop) {
+      auto txrespptr(std::move(qu.front()));
+      qu.pop();
+      return {txrespptr, elapsed};
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:209]:
```c++
return { std::move(txrespptr), elapsed };

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 12]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+    if (qu.empty()) {
+      while (
+          std::chrono::steady_clock::now() < deadline
+          and not cv.wait_until(lk, deadline, [&] { return not qu.empty(); }))
+        ;
+    }
+    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+        deadline - std::chrono::steady_clock::now());
+    if (qu.empty())  // timed out and still empty
+      return {nullptr, elapsed};
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:205]: format plz

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 13]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+  // std::chrono::milliseconds timeout;
+  std::mutex mtx;
+  std::condition_variable cv;
+  std::unordered_map<HashType, std::queue<TxResponsePtr>, HashType::Hasher> map;
+
+  template <bool do_pop = false>
+  auto wait_get(HashType const &txhash, std::chrono::milliseconds timeout)
+      -> WaitGetResult {  // std::tuple<TxResponsePtr,
+                          // std::chrono::milliseconds> {  //
+                          // std::optional<TxResponsePtr>
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:188]: remove comment plz

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 14]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+    if (qu.empty())  // timed out and still empty
+      return {nullptr, elapsed};
+    if constexpr (do_pop) {
+      auto txrespptr(std::move(qu.front()));
+      qu.pop();
+      return {txrespptr, elapsed};
+    } else {
+      auto txrespptr = qu.front();
+      return {txrespptr, elapsed};
+      ;
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:213]: remove plz

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 15]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+        deadline - std::chrono::steady_clock::now());
+    if (qu.empty())  // timed out and still empty
+      return {nullptr, elapsed};
+    if constexpr (do_pop) {
+      auto txrespptr(std::move(qu.front()));
+      qu.pop();
+      return {txrespptr, elapsed};
+    } else {
+      auto txrespptr = qu.front();
+      return {txrespptr, elapsed};
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:212]:
```c++
return { qu.front(), elapsed };

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 16]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+      ;
+    }
+  }
+
+ public:
+  // ResponsesQueues(std::chrono::milliseconds ms) : timeout(ms) {}
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:218]: remove plz

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 17]
```diff
@@ -113,46 +109,132 @@ using integration_framework::IntegrationTestFramework;
...
+  void push(TxResponsePtr p_txresp) {
+    assert(p_txresp);
+    std::unique_lock lk(mtx);
+    auto it = map.try_emplace(p_txresp->transactionHash()).first;
+    it->second.push(std::move(p_txresp));
```
Comment (iceseer) [inline test/framework/integration_framework/integration_test_framework.cpp:223]:
```c++
map[p_txresp->transactionHash()].push(std::move(p_txresp));

**Review by iceseer — APPROVED** [review]

File: test/framework/integration_framework/fake_peer/fake_peer.cpp (hunk 8 above)
Comment (kuvaldini) [reply test/framework/integration_framework/fake_peer/fake_peer.cpp]: removed

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 10 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.cpp]: done

File: test/framework/integration_framework/integration_test_framework.cpp (hunk 17 above)
Comment (kuvaldini) [reply test/framework/integration_framework/integration_test_framework.cpp]: doen

File: irohad/main/server_runner.cpp [hunk 18]
```diff
@@ -59,10 +57,7 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
   grpc::ServerBuilder builder;
   int selected_port = 0;
 
-  if (not reuse_) {
-    builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, 0);
-  }
-
+  builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, reuse_?1:0);
```
Comment (kamilsa) [inline irohad/main/server_runner.cpp:60]: is it clang formatted?

File: test/framework/integration_framework/fake_peer/fake_peer.hpp [hunk 19]
```diff
@@ -98,11 +109,15 @@ namespace integration_framework::fake_peer {
     ProposalStorage &getProposalStorage();
 
     /// Start the fake peer.
-    std::unique_ptr<iroha::network::ServerRunner> run();
+    std::unique_ptr<iroha::network::ServerRunner> run(bool reuse_port = false);
```
Comment (kamilsa) [inline test/framework/integration_framework/fake_peer/fake_peer.hpp:112]: I would add comment explaining reasoning of new parameter. Probably worth mentioning that was added to run tests in parallel?

**Review by kamilsa — APPROVED** [review]

File: test/framework/integration_framework/fake_peer/fake_peer.hpp (hunk 19 above)
Comment (kuvaldini) [reply test/framework/integration_framework/fake_peer/fake_peer.hpp:113]: commented

**Pull request merged**